File: touchfish-core/src/domain/fish.rs
use serde::{Deserialize, Serialize};
use serde_with::skip_serializing_none;
use strum_macros::{Display, EnumString};
use yfunc_rust::{prelude::*, YBytes, YTime};

#[yfunc]
#[derive(Serialize, Debug)]
pub struct Fish {
    pub identity: String,
    pub count: i32,
    pub fish_type: FishType,
    pub fish_data: YBytes,
    pub data_info: DataInfo,
    pub desc: String,
    pub tags: Vec<String>,
    pub is_marked: bool,
    pub is_locked: bool,
    pub extra_info: String,
    pub create_time: YTime,
    pub update_time: YTime,
}

impl Fish {

    pub fn to_preview_json(&self) -> YRes<String> {
        FishPreview::from_fish(self).trace(
            ctx!("serialize Fish to preview style json string -> build FishPreview from Fish: FishPreview::from_fish failed", self.identity)
        )?.to_pretty_json_str().trace(
            ctx!("serialize Fish to preview style json string -> serialize FishPreview to json: FishPreview::to_pretty_json_str failed", self.identity)
        )
    }

}

#[yfunc]
#[derive(Serialize, Debug)]
pub struct FishPreview {
    pub identity: String,
    pub count: i32,
    pub fish_type: FishType,
    pub data_preview: Option<String>,
    pub data_info: DataInfo,
    pub desc: String,
    pub tags: Vec<String>,
    pub is_marked: bool,
    pub is_locked: bool,
    pub extra_info: String,
    pub create_time: String,
    pub update_time: String,
}

impl FishPreview {

    pub fn from_fish(fish: &Fish) -> YRes<FishPreview> {
        let data_preview = match fish.fish_type {
            FishType::Text => {
                let preview = fish.fish_data.to_str().trace(
                    ctx!("build FishPreview by Fish -> parse fish_data to text for text type fish: fish.fish_data.to_str() failed", fish.identity)
                )?;
                Some(preview.chars().take(80).collect())
            },
            _ => None,
        };
        let create_time = fish.create_time.east8().trace(
            ctx!("build FishPreview by Fish -> get east8 time string of create_time: fish.create_time.east8() failed", fish.identity)
        )?;
        let update_time = fish.update_time.east8().trace(
            ctx!("build FishPreview by Fish -> get east8 time string of update_time: fish.update_time.east8() failed", fish.identity)
        )?;
        Ok(FishPreview { 
            identity: fish.identity.clone(), count: fish.count, fish_type: fish.fish_type,
            data_preview, data_info: fish.data_info.clone(), desc: fish.desc.clone(), tags: fish.tags.clone(),
            is_marked: fish.is_marked, is_locked: fish.is_locked, extra_info: fish.extra_info.clone(),
            create_time, update_time,
        })
    }

}

#[yfunc]
#[derive(Serialize, Deserialize, Debug, EnumString, Display, PartialEq, Eq, Hash, Clone, Copy)]
pub enum FishType {
    Text,
    Image,
    Other,
}

#[yfunc]
#[skip_serializing_none]
#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct DataInfo {
    pub byte_count: Option<usize>,
    // Text
    pub char_count: Option<usize>,
    pub word_count: Option<usize>,
    pub row_count: Option<usize>,
    // Image
    pub width: Option<usize>,
    pub height: Option<usize>,
}

impl DataInfo {

    pub fn new() -> DataInfo {
        DataInfo { 
            byte_count: None, 
            char_count: None, word_count: None, row_count: None,
            width: None, height: None,
        }
    }

}

File: touchfish-macos-client/TouchFish/Fish/Fish.swift
import AppKit
import SwiftUI

class Fish {
    
    enum FishType: String, CaseIterable {
        case Text
        case Image
        case Other
    }
    
    struct DataInfo: Codable {
        let byteCount: Int?
        let charCount: Int?
        let wordCount: Int?
        let rowCount: Int?
        let width: Int?
        let height: Int?
    }
    
    let uid: String
    let identity: String
    let fishType: FishType
    let fishData: Data
    let dataInfo: DataInfo
    let description: String
    let tags: [String]
    let isMarked: Bool
    let isLocked: Bool
    let extraInfo: [String:String]
    let createTime: String
    let updateTime: String
    
    let textData: String?
    let imageData: NSImage?
    
    init(
        uid: String, identity: String, fishType: FishType, fishData: Data,
        dataInfo: DataInfo, description: String, tags: [String], isMarked: Bool, isLocked: Bool,
        extraInfo: [String:String], createTime: String, updateTime: String
    ) {
        self.uid = uid
        self.identity = identity
        self.fishType = fishType
        self.fishData = fishData
        self.dataInfo = dataInfo
        self.description = description
        self.tags = tags
        self.isMarked = isMarked
        self.isLocked = isLocked
        self.extraInfo = extraInfo
        self.createTime = createTime
        self.updateTime = updateTime
        var textData: String? = nil
        var imageData: NSImage? = nil
        switch fishType {
        case .Text:
            textData = String(data: fishData, encoding: .utf8)
        case .Image:
            imageData = NSImage(data: fishData)
        default:
            break
        }
        self.textData = textData
        self.imageData = imageData
    }
    
    func withMark(isMarked: Bool) -> Fish {
        Fish(
            uid: self.uid, identity: self.identity, fishType: self.fishType, fishData: self.fishData, dataInfo: self.dataInfo,
            description: self.description, tags: self.tags, isMarked: isMarked, isLocked: self.isLocked, extraInfo: self.extraInfo,
            createTime: self.createTime, updateTime: self.updateTime
        )
    }
    
    func withLock(isLocked: Bool) -> Fish {
        Fish(
            uid: self.uid, identity: self.identity, fishType: self.fishType, fishData: self.fishData, dataInfo: self.dataInfo,
            description: self.description, tags: self.tags, isMarked: self.isMarked, isLocked: isLocked, extraInfo: self.extraInfo,
            createTime: self.createTime, updateTime: self.updateTime
        )
    }
    
    var defaultLinePreview: String {
        return "\(self.fishType.rawValue):\(self.uid)"
    }
    
    var linePreview: String {
        var ret: String = self.defaultLinePreview
        switch self.fishType {
        case .Text:
            if self.description.count > 0 {
                ret = Functions.getLinePreview(self.description)
                break
            }
            if let textData = self.textData {
                ret = Functions.getLinePreview(textData)
            }
        case .Image:
            if self.description.count > 0 {
                ret = Functions.getLinePreview(self.description)
            }
        default:
            if self.description.count > 0 {
                ret = Functions.getLinePreview(self.description)
            }
        }
        return ret
    }
    
    var fishIcon: Image {
        switch self.fishType {
        case .Text:
            return Image(systemName: "doc.plaintext")
        case .Image:
            return Image(systemName: "photo")
        default:
            return Image(systemName: "fish")
        }
    }
    
    func copyToClipboard() {
        Functions.copyDataToClipboard(data: self.fishData, type: self.fishType)
    }
    
}
